In [7]:
pip install pandas

  Using cached pandas-3.0.5-cp312-cp312-macosx_11_0_arm64.whl.metadata (79 kB)
Using cached pandas-3.0.5-cp312-cp312-macosx_11_0_arm64.whl (10.0 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.4/5.4 MB 11.4 MB/s  0:00:00eta 0:00:01m
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2/2 [pandas]2m1/2 [pandas]
Note: you may need to restart the kernel to use updated packages.


In [ ]:
import pandas as pd 



In [10]:
pip install openpyxl

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2/2 [openpyxl]
Note: you may need to restart the kernel to use updated packages.


In [11]:
import openpyxl

In [12]:
file_path = "../data/raw/company_sales.xlsx"

# Read both sheets from the Excel workbook.
sales = pd.read_excel(file_path, sheet_name="Sales Data")
people = pd.read_excel(file_path, sheet_name="People")

In [13]:
# Check how many rows and columns each table has.
print("Sales shape:", sales.shape)
print("People shape:", people.shape)

Sales shape: (435, 10)
People shape: (9, 2)


In [14]:
# Print the column names so we know exactly what information we have.
print("Sales columns:")
print(sales.columns.tolist())

print("\nPeople columns:")
print(people.columns.tolist())

Sales columns:
['Opportunity ID', 'Opp Owner User ID', 'Opp Owner Geo', 'Segment', 'Opportunity Type', 'Close Date', 'Product Line', 'Product Family', 'Total Price', 'Rep Name']

People columns:
['Opp Owner User ID', 'Name']


In [15]:
# Show the first 5 rows of the Sales table.
sales.head()

,Opportunity ID,Opp Owner User ID,Opp Owner Geo,Segment,Opportunity Type,Close Date,Product Line,Product Family,Total Price,Rep Name
0,0061Y00000QA0nIxRU,0054z00000AAVA1E0V,AMER,Enterprise,New Business,2019-09-06,Workforce,Workforce - C,48989.32,Alice Chen
1,0061Y00000QA0nIxRU,0054z00000AAVA1E0V,AMER,Enterprise,New Business,2019-09-06,Workforce,Workforce - B,36741.99,Alice Chen
2,0061Y00000QA0nIxRU,0054z00000AAVA1E0V,AMER,Enterprise,New Business,2019-09-06,Workforce,Workforce - A,24494.66,Alice Chen
3,0061Y00000QA0nIxRU,0054z00000AAVA1E0V,AMER,Enterprise,New Business,2019-09-06,Workforce,Workforce - A,24494.66,Alice Chen
4,0061Y00000QA0nIxXH,0054z00000AAFA1E0p,AMER,Enterprise,Upsell,2019-04-12,Workforce,Workforce - B,8461.79,Bob Smith


In [16]:
# Show the first 5 rows of the People table.
people.head()

,Opp Owner User ID,Name
0,0054z00000AAVA1E0V,Alice Chen
1,0054z00000AAFA1E0p,Bob Smith
2,0051Y00000QAVBgScR,Charlie Day
3,0054z00000AAJA0qTj,Diana Prince
4,0054z00000AAFA1E0i,Edward Nygma


In [17]:
# Check what type of data Python thinks each column contains.
sales.info()

<class 'pandas.DataFrame'>
RangeIndex: 435 entries, 0 to 434
Data columns (total 10 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   Opportunity ID     435 non-null    str           
 1   Opp Owner User ID  435 non-null    str           
 2   Opp Owner Geo      435 non-null    str           
 3   Segment            435 non-null    str           
 4   Opportunity Type   435 non-null    str           
 5   Close Date         435 non-null    datetime64[us]
 6   Product Line       435 non-null    str           
 7   Product Family     435 non-null    str           
 8   Total Price        435 non-null    float64       
 9   Rep Name           113 non-null    str           
dtypes: datetime64[us](1), float64(1), str(8)
memory usage: 34.1 KB


In [18]:
# Count how many missing values exist in each Sales column.
sales.isna().sum()

Opportunity ID         0
Opp Owner User ID      0
Opp Owner Geo          0
Segment                0
Opportunity Type       0
Close Date             0
Product Line           0
Product Family         0
Total Price            0
Rep Name             322
dtype: int64

In [19]:
# Count missing values in the People table.
people.isna().sum()

Opp Owner User ID    0
Name                 0
dtype: int64

In [20]:
# Show the different segments in the dataset.
print("Segments:")
print(sales["Segment"].unique())

# Show the different opportunity types.
print("\nOpportunity Types:")
print(sales["Opportunity Type"].unique())

# Show the different geographies.
print("\nGeographies:")
print(sales["Opp Owner Geo"].unique())

# Show the product lines.
print("\nProduct Lines:")
print(sales["Product Line"].unique())

# Show the product families.
print("\nProduct Families:")
print(sales["Product Family"].unique())

Segments:
<StringArray>
['Enterprise', 'Corporate']
Length: 2, dtype: str

Opportunity Types:
<StringArray>
['New Business', 'Upsell']
Length: 2, dtype: str

Geographies:
<StringArray>
['AMER', 'EMEA', 'APAC']
Length: 3, dtype: str

Product Lines:
<StringArray>
['Workforce', 'CIAM']
Length: 2, dtype: str

Product Families:
<StringArray>
['Workforce - C', 'Workforce - B', 'Workforce - A',      'CIAM - A',
 'Workforce - D', 'Workforce - E']
Length: 6, dtype: str


In [21]:
# Count the total number of rows.
total_rows = len(sales)

# Count how many different opportunities exist.
unique_opportunities = sales["Opportunity ID"].nunique()

print("Total rows:", total_rows)
print("Unique opportunities:", unique_opportunities)

Total rows: 435
Unique opportunities: 145


In [22]:
# Count how many product-family rows belong to each opportunity.
rows_per_opportunity = sales.groupby("Opportunity ID").size()

# Show the first few results.
rows_per_opportunity.head()

Opportunity ID
0061Y00000QA0nIxRU    4
0061Y00000QA0nIxXH    3
0061Y00000QA0nJ1PR    5
0061Y00000QA0nJ6NX    1
0061Y00000QA0nUpUC    3
dtype: int64

In [23]:
# Understand how many rows opportunities usually contain.
rows_per_opportunity.describe()

count    145.000000
mean       3.000000
std        1.348868
min        1.000000
25%        2.000000
50%        3.000000
75%        4.000000
max        6.000000
dtype: float64

In [24]:
# Find rows where every column is exactly duplicated.
duplicate_rows = sales.duplicated()

print("Number of exact duplicate rows:", duplicate_rows.sum())

Number of exact duplicate rows: 52


In [25]:
# Find all owner IDs that appear in Sales Data.
sales_owner_ids = set(sales["Opp Owner User ID"].dropna())

# Find all owner IDs that appear in the People table.
people_owner_ids = set(people["Opp Owner User ID"].dropna())

# Find Sales owners that don't have a matching person in the People table.
unmatched_owner_ids = sales_owner_ids - people_owner_ids

print("Sales owner IDs:", len(sales_owner_ids))
print("People owner IDs:", len(people_owner_ids))
print("Unmatched owner IDs:", len(unmatched_owner_ids))

Sales owner IDs: 61
People owner IDs: 9
Unmatched owner IDs: 52


In [26]:
# Convert Close Date to a real datetime format.
sales["Close Date"] = pd.to_datetime(sales["Close Date"])

# Find the earliest and latest opportunity dates.
print("Earliest close date:", sales["Close Date"].min())
print("Latest close date:", sales["Close Date"].max())

Earliest close date: 2019-03-14 00:00:00
Latest close date: 2022-01-20 00:00:00


In [ ]:
# Check the data type of Total Price.
print(sales["Total Price"].dtype)

# Show a few price values.  
print(sales["Total Price"].head())

float64
0    48989.32
1    36741.99
2    24494.66
3    24494.66
4     8461.79
Name: Total Price, dtype: float64


### Data Quality & Initial Validation

The analysis began by validating the structure, completeness, and quality of the raw Sales and People datasets. The Sales Data contains 435 records across 10 columns, while the People table contains 9 records across 2 columns. There are 145 unique opportunities, confirming that individual opportunities can contain multiple product-family records. The dataset contains 52 exact duplicate records that required further investigation. The Sales Data contains 61 unique owner IDs, while only 9 are mapped in the People table, resulting in substantial missing representative-name information. No missing values were identified in the primary Sales Data fields aside from the incomplete owner-name mapping. Close Date is stored as a valid datetime covering March 2019 through January 2022, and Total Price is stored as a numeric field. The categorical variables include 3 owner geographies, 2 segments, 2 opportunity types, 2 product lines, and 6 product families. These checks established the primary data-quality considerations and confirmed that the dataset could proceed to structured cleaning and analytical modeling.